x-pack/plugins/apm/server/lib/services/get_service_transaction_groups/index.ts

```diff
@@ -0,0 +1,322 @@
+  getTransactionDurationFieldForAggregatedTransactions,
+} from '../../helpers/aggregated_transactions';
+
+export type ServiceTransactionGroupItem = ValuesType<
+  PromiseReturnType<typeof getServiceTransactionGroups>['transaction_groups']
+>;
+
+export async function getServiceTransactionGroups({
+  serviceName,
+  setup,
+  size,
+  numBuckets,
+  pageIndex,
+  sortDirection,
+  sortField,
+  searchAggregatedTransactions,
+}: {
+  serviceName: string;
+  setup: Setup & SetupTimeRange;
+  size: number;
+  pageIndex: number;
+  numBuckets: number;
+  sortDirection: 'asc' | 'desc';
+  sortField: 'latency' | 'traffic' | 'error_rate' | 'impact';
+  searchAggregatedTransactions: boolean;
+}) {
+  const { apmEventClient, start, end, esFilter } = setup;
+
+  const { intervalString } = getBucketSize(start, end, numBuckets);
+
+  const response = await apmEventClient.search({
+    apm: {
+      events: [
+        getProcessorEventForAggregatedTransactions(
+          searchAggregatedTransactions
+        ),
+      ],
+    },
+    body: {
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            { term: { [SERVICE_NAME]: serviceName } },
+            { range: rangeFilter(start, end) },
+            ...esFilter,
+          ],
+        },
+      },
+      aggs: {
+        transaction_groups: {
+          terms: {
+            field: TRANSACTION_NAME,
+            size: 500,
+            order: {
+              _count: 'desc',
+            },
+          },
+          aggs: {
+            avg_latency: {
+              avg: {
+                field: getTransactionDurationFieldForAggregatedTransactions(
+                  searchAggregatedTransactions
+                ),
+              },
+            },
+            transaction_count: {
+              value_count: {
+                field: getTransactionDurationFieldForAggregatedTransactions(
+                  searchAggregatedTransactions
+                ),
+              },
+            },
+            [EVENT_OUTCOME]: {
+              terms: {
+                field: EVENT_OUTCOME,
+              },
+              aggs: {
+                transaction_count: {
+                  value_count: {
+                    field: getTransactionDurationFieldForAggregatedTransactions(
+                      searchAggregatedTransactions
+                    ),
+                  },
+                },
+              },
+            },
+          },
+        },
+      },
+    },
+  });
+
+  const transactionGroups =
+    response.aggregations?.transaction_groups.buckets.map((bucket) => ({
+      name: bucket.key as string,
+      latency: bucket.avg_latency.value,
+      traffic: bucket.transaction_count.value,
+      error_rate:
+        bucket.transaction_count.value > 0
+          ? (bucket[EVENT_OUTCOME].buckets.find(
+              (outcomeBucket) => outcomeBucket.key === EventOutcome.failure
+            )?.transaction_count.value ?? 0) / bucket.transaction_count.value
+          : null,
```

Comment:
I find this slightly hard to read but maybe just me 😅

Reply:
I can see how it's hard to read. Will have a look on how to improve. It's a little hard because the fastest query is not necessarily the most ideal data structure. I still need to figure out what is faster, I might try a filter agg or a terms agg with a terms filter.